File: backend/semgrep_rules/tests/python/kisa_sw_12_ssrf.py
import httpx
import requests
import urllib.request


def unsafe_requests_get(request):
    url = request.GET.get("url")

    # ruleid: kisa.sw12.python.external-input-to-http-client
    return requests.get(url)


def unsafe_requests_method(request):
    url = request.POST["callback_url"]
    method = request.POST.get("method")

    # ruleid: kisa.sw12.python.external-input-to-http-client
    return requests.request(method, url, timeout=5)


def unsafe_httpx_multistep(request):
    target = request.headers.get("X-Target-URL")
    copied_target = target

    # ruleid: kisa.sw12.python.external-input-to-http-client
    return httpx.post(copied_target, json={"status": "ready"})


def unsafe_urllib(request):
    target = request.args["url"]

    # ruleid: kisa.sw12.python.external-input-to-http-client
    return urllib.request.urlopen(target, timeout=5)


def safe_hardcoded_requests():
    # ok: kisa.sw12.python.external-input-to-http-client
    return requests.get("https://api.example.com/status")


def safe_hardcoded_httpx():
    # ok: kisa.sw12.python.external-input-to-http-client
    return httpx.get("https://api.example.com/status")

File: backend/semgrep_rules/tests/javascript/kisa_sw_12_ssrf.js
const axios = require("axios");
const http = require("http");
const https = require("https");
const undici = require("undici");


function unsafeAxiosGet(req) {
  const url = req.query.url;

  // ruleid: kisa.sw12.javascript.external-input-to-http-client
  return axios.get(url);
}


function unsafeAxiosMethod(req) {
  const url = req.body["callbackUrl"];

  // ruleid: kisa.sw12.javascript.external-input-to-http-client
  return axios.post(url, { status: "ready" });
}


function unsafeFetchMultistep(req) {
  const target = req.params.url;
  const copiedTarget = target;

  // ruleid: kisa.sw12.javascript.external-input-to-http-client
  return fetch(copiedTarget);
}


function unsafeNodeHttp(req) {
  const target = req.headers["x-target-url"];

  // ruleid: kisa.sw12.javascript.external-input-to-http-client
  return http.request(target);
}


function unsafeNodeHttps(req) {
  const target = req.get("X-Target-URL");

  // ruleid: kisa.sw12.javascript.external-input-to-http-client
  return https.get(target);
}


function unsafeUndici(req) {
  const target = req.query["target"];

  // ruleid: kisa.sw12.javascript.external-input-to-http-client
  return undici.request(target);
}


function safeHardcodedAxios() {
  // ok: kisa.sw12.javascript.external-input-to-http-client
  return axios.get("https://api.example.com/status");
}


function safeHardcodedFetch() {
  // ok: kisa.sw12.javascript.external-input-to-http-client
  return fetch("https://api.example.com/status");
}
